ouroboros-consensus-cardano-tools/src/Cardano/Tools/ImmDBServer/MiniProtocols.hs

```diff
@@ -0,0 +1,229 @@
+    iterator <- ImmutableDB.streamAll immDB registry blockComponent
+    runChainSyncServer $ idle iterator Nothing
+  where
+    idle ::
+         Iterator m blk a
+      -> Maybe (Point blk) -- just-negotiated intersection, if any
+      -> ChainSyncServer a (Point blk) (Tip blk) m ()
+    idle iterator mIntersectionPt =
+        ChainSyncServer $ pure ServerStIdle {
+            recvMsgRequestNext   = handleRequestNext'
+          , recvMsgFindIntersect = handleFindIntersect iterator
+          , recvMsgDoneClient    = pure ()
+          }
+      where
+        handleRequestNext' = case mIntersectionPt of
+          Just pt -> do
+            tip <- getImmutableTip
+            pure $ Left $ SendMsgRollBackward pt tip $ idle iterator Nothing
+          Nothing -> handleRequestNext iterator
+
+    handleRequestNext ::
+         Iterator m blk a
+      -> m (Either (ServerStNext a (Point blk) (Tip blk) m ())
+                (m (ServerStNext a (Point blk) (Tip blk) m ())))
+    handleRequestNext iterator =
+       ImmutableDB.iteratorNext iterator >>= \case
+         ImmutableDB.IteratorExhausted ->
+           throwIO ReachedImmutableTip
+         ImmutableDB.IteratorResult a  -> Left <$> do
+           tip <- getImmutableTip
+           pure $ SendMsgRollForward a tip $ idle iterator Nothing
+
+    handleFindIntersect ::
+         Iterator m blk a
+      -> [Point blk]
+      -> m (ServerStIntersect a (Point blk) (Tip blk) m ())
+    handleFindIntersect iterator = go
+      where
+        go (pt : pts) =
+          ImmutableDB.streamAfterPoint immDB registry blockComponent pt >>= \case
+            Left _          -> go pts
+            Right iterator' -> do
+              ImmutableDB.iteratorClose iterator
+              tip <- getImmutableTip
+              pure $ SendMsgIntersectFound pt tip $ idle iterator' (Just pt)
+        go [] = do
+          tip <- getImmutableTip
+          pure $ SendMsgIntersectNotFound tip $ idle iterator Nothing
+
+    getImmutableTip :: m (Tip blk)
+    getImmutableTip = atomically (ImmutableDB.getTip immDB) <&> \case
+        Origin                         -> TipGenesis
+        NotOrigin ImmutableDB.Tip {..} -> Tip tipSlotNo tipHash tipBlockNo
+
+blockFetchServer ::
+     forall m blk a. (IOLike m)
+  => ImmutableDB m blk
+  -> BlockComponent blk a
+  -> ResourceRegistry m
+  -> BlockFetchServer a (Point blk) m ()
+blockFetchServer immDB blockComponent registry = idle
+  where
+    idle :: BlockFetchServer a (Point blk) m ()
+    idle = flip BlockFetchServer () \(ChainRange from to) -> do
+        streamTo   <- StreamFromInclusive <$> toRealPoint from
+        streamFrom <- StreamToInclusive <$> toRealPoint to
+        ImmutableDB.stream immDB registry blockComponent streamTo streamFrom <&> \case
+          Left _         -> SendMsgNoBlocks $ pure idle
+          Right iterator -> SendMsgStartBatch $ sendBlocks iterator
+      where
+        toRealPoint :: Point blk -> m (RealPoint blk)
```

Comment:
Suggestion: I also had to implement this one, shouldn't we upstream it (without having it partial)?

Reply:
What exactly would you like to upstream in addition to [`pointToWithOriginRealPoint`](https://input-output-hk.github.io/ouroboros-network/ouroboros-consensus/Ouroboros-Consensus-Block-RealPoint.html#v:pointToWithOriginRealPoint) that is used here?

Reply:
Probably nothing, it's too simple, but I find it "smelly" both of us do the exact same transformation, for the same reason, a few days apart.